### ISSUE TITLE
Documentation generator does not like Python 3

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1186]
The documentation generator does not like Python 3. It would be useful to update it, for convenience.

Terminal transcript:

```
machine:~/Repos/igraph-c/igraph/doc $ make html
../doc/doxrox.py -t vector.xxml -e ../doc/c-docbook.re -o vector.xml ../include/igraph_vector_pmt.h \
	../src/vector.pmt ../src/vector_ptr.c ../src/vector.c \
	../include/igraph_vector_ptr.h
  File "../doc/doxrox.py", line 93
    print 'Reading regular expressions...',
                                         ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print('Reading regular expressions...', end=" ")?
make: *** [vector.xml] Error 1
machine:~/Repos/igraph-c/igraph/doc $ 
```

While the changes that are needed are likely trivial, I'd rather leave this to someone more experienced in Python than me.

**Comment by iosonofabio** [comment]
Happy to take this one, done this kinda thing many times

**Comment by iosonofabio** [comment]
Hmm I fixed the `print` easily enough, but now there is a mess with regular expressions and escape characters:

```python
../doc/doxrox.py -t vector.xxml -e ../doc/c-docbook.re -o vector.xml ../include/igraph_vector_pmt.h \
../src/vector.pmt ../src/vector_ptr.c ../src/vector.c \
../include/igraph_vector_ptr.h
Traceback (most recent call last):
  File "/usr/lib/python3.7/sre_parse.py", line 1021, in parse_template
    this = chr(ESCAPES[this][1])
KeyError: '\\p'
```

Seems like there is a script called `doxrox.py` that is undocumented and does some kind of regex black magic. It sources "conversions" from another file called `c-docbook.re` with some regex language spec interspersed with ancient Python code.

Anyone knows the rationale of this? Injecting regex into Python and Python into regex source definitions seems like a pretty evil plan...

**Comment by ntamas** [comment]
Well, yes, `doxrox.py` is a hack. The whole store is that we initially used Doxygen for documentation until we realized that we needed to support Docbook (so we can generate documentation in HTML, PDF and Texinfo, all from the same sources). At that point it was easier to write a script that "parses" Doxygen docstrings with regexes and converts them into Docbook format. The rules for this conversion are stored in `c-docbook.re`.

I can take a closer look at it if you wish.

**Comment by iosonofabio** [comment]
Thank you for the explanation, I am going through it for now, will let you know if that efforts fails :-)

**Comment by iosonofabio** [comment]
I think I'm getting this to work, but it is going to be impossible to maintain. In my corner of the programming world one would use Doxygen + Sphinx + Breathe:

https://breathe.readthedocs.io/en/latest/

Which does the same except we don't have to maintain it ourselves. Would be great to hear people's comments on this as opposed to Docbook. Or we could just drop texinfo support, if people are not too opinionated about that.

**Comment by ntamas** [comment]
> I think I'm getting this to work, but it is going to be impossible to maintain.
Yeah, let's fix it and then treat it as a black box from then on :) Eventually we should switch to a better solution but I'm not sure what that will be. Even though we started from the Doxygen syntax, if you look at the docstrings you'll see that we have deviated towards a strange mixture of Doxygen + Docbook. I think that in the long run we should revert back to plain Doxygen and drop support for PDF + Texinfo (unless @gaborcsardi needs it for the R interface).

FWIW I like [Antora](https://antora.org/) a lot for writing technical documentation - if only we could get Sphinx-generated API docs into Antora somehow.

**Comment by iosonofabio** [comment]
Hi all,

I've been trying all kinds of things for two days and am going to abandon it, sorry about that. Let me know if anyone wants my playground code to try themselves.

I would like to share some issues I encountered in case you are willing to give it a go:
- adding brackets to `print` statements is a breeze of course
- escaping changed in Python 3 (due to unicode) and again in Python 3.6 (deprecation of incorrect but extremely widespread patterns) - especialy `\\p` is tricky
- the docs are mix of doxygen and docbook, and at least one formatting error was found in `about_igraph_vector_t_objects`
- the rules in `c-docbook.re` are not really explained, and `#`-style comments are not actually supported by the parser, so you have to fix the parsing of that file first. Same goes for newlines in the middle of rules, you probably want to trash those
- there is no working version `doxrox.py`, so I can't check what the output is supposed to be
- injected python code in `c-docbook.re` is going to be difficult to debug at best, I didn't even get there
- at least three regex rules rely on the `re.MULTILINE` heavily to work, since they use the ambiguous `$`pattern to mean `\n`

I think keeping a mix of doxygen and docbook and maintaining a whole cross-language doc converter based on regex is bad enough that we should abandon it immediately and switch to another solution (see below). I understand the wish "let's fix it now, then refactor later" but I sincerely could not even get close to fixing this, so I'm afraid that plan does not work here ;-)

Instead, I propose:
- we switch all the docbook chunks of the inline docs back to doxygen
- we dump texinfo for now and use doxygen to generate html and PDF code
- we start exploring extensions using sphinx, antora, you name it. This step will likely happen at a later stage.

What do you guys think?

**Comment by iosonofabio** [comment]
I'm looking at the output of `make html` when hacked to force Python2. Although in that case at least the output `xml` files make some sense when inspected by eye, the command eventually fails because of non-compliant XML format:

```bash
vector.xml:2428: element link: validity error : IDREF attribute linkend references an unknown ID "igraph_decompose"
tutorial.xml:115: element link: validity error : IDREF attribute linkend references an unknown ID "igraph_shortest_paths"
tutorial.xml:225: element link: validity error : IDREF attribute linkend references an unknown ID "igraph_simplify"
pmt.xml:116: element link: validity error : IDREF attribute linkend references an unknown ID "MATRIX"
Document /home/fabio/university/PI/projects/igraph/doc/igraph-docs.xml does not validate
make: *** [Makefile:667: html/stamp] Error 13
```

So I am wondering whether the docs are currently building on anyone's system? or is it currently broken even in Python2?

Thanks!

**Comment by iosonofabio** [comment]
closed in favor of #1188.

**Comment by szhorvat** [comment]
> So I am wondering whether the docs are currently building on anyone's system? or is it currently broken even in Python2?

Yes, I build them often. I built them today. I need to do this to check that any edits I make to the docs are fine.

**Comment by iosonofabio** [comment]
@szhorvat Oh I closed the issue at the same time. Would you like to reopen it? In that case I think it may be best if you take it, since on my system the resulting xml files do not validate...

**Comment by szhorvat** [comment]
I use it (because I write docs), but that doesn't mean I know how it works.

Since this is difficult to fix, but there is a very easy workaround (just use Python 2.7), I think this should be of low priority. It's not a user-facing thing, it's only for developers. IMO we don't need to fix this before 0.8.

**Comment by iosonofabio** [comment]
ok, thanks for the explanation. I'll leave it closed then. Since it's long term, we might as well aim for a more maintainable system (see #1188).
